### ISSUE TITLE
Error with cmake:  Prevented in-tree built.  Please create a build directory outside of the SDL source code and call cmake from there

### ISSUE THREAD

**Issue by rondadon** [ptitSeb/SDL2#1]
Hi there!

I just wanted to build SDL2 from your GIT Sources but as I tried to "cmake" it I ran into an issue which prevents me from building SDL2 for my RPI 3B.

I followed your instructions at: https://magazine.odroid.com/article/playing-modern-fna-games-on-the-odroid-platform/ .

After creating the "build" folder in the SDL2 Folder and "cd-ing" into the "build" folder and executing the command "cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo." i get the following error message:

`CMake Error at CMakeLists.txt:4 (message):
  Prevented in-tree built.  Please create a build directory outside of the
  SDL source code and call cmake from there
-- Configuring incomplete, errors occurred!`

in the parent Folder SDL2 there is a "CMakeLists.txt" which contains following code:

`if(${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
 message(FATAL_ERROR "Prevented in-tree built. Please create a build directory outside of the SDL source code and call cmake from there")
endif()`

So this is the code which is called but I followed your instruction and still get this error. Even tho I am in the subfolder "build" inside the SDL2 Folder when calling the following command:

` cmake .. -DCMAKE_BUILD_TYPE=RelWithDebInfo.`

So may I ask you how we/you/I can fix this problem? Can you reproduce this issue?

Thanks in advance and best regards :) 👍 
kama

**Comment by ptitSeb** [comment]
Mmm, I'm surprised you have this, because this is how I built currently.
Well, I can see now that there is, of course, a typo in the article, and the `..` is indeed missing from the command line :( .
Did you tried to built it in-source first before creating `build` ? If yes, cmake then created already so files that should be removed (or, easier way, remove the whole SDL2 folder and clone it again).
If no, that strange, I'll retest on my side later. In the mean time, you can alway create a build folder out of the sdl2 folder and try again. Something like:
```BASH
cd ~
git clone https://github.com/ptitSeb/SDL2.git
mkdir sdl2_build
cd sdl2_build
cmake ../SDL2 -DCMAKE_BUILD_TYPE=RelWithDebInfo.
```
Should do the trick.

**Comment by ZangLucarson** [comment]
Hey, I tried this version of the compilation command and ended up with this error. CMake Error at cmake/macros.cmake:28 (message):
```
  *** ERROR: Missing Xext.h, maybe you need to install the libxext-dev
  package?
Call Stack (most recent call first):
  cmake/sdlchecks.cmake:413 (message_error)
  CMakeLists.txt:1043 (CheckX11)
```
Any clue what the issue is?

**Comment by ptitSeb** [comment]
Just what the message said: you are missing some X11 header.
On debian and friends, try `sudo apt install libxext-dev`

**Comment by ZangLucarson** [comment]
> Just what the message said: you are missing some X11 header.
> On debian and friends, try `sudo apt install libxext-dev`
I just installed libxext-dev as you said, I'm still getting the same error.
Would I have to get a fresh copy of SDL2?

**Comment by ZangLucarson** [comment]
Thank you!

**Comment by ZangLucarson** [comment]
This is more related to the FNA games article, but I'm having an issue while compile mojoshader. In the article it says to copy "libmojoshader.so" but when I try to copy it, the terminal says that the file does not exist. Any clue as to how I can fix that?

**Comment by ptitSeb** [comment]
There must have been an error during the build, the "libmojoshader.so" should be created at the base of the mojoshader folder.
Try do check with `ls -l` the content of the mojoshader folder also.

**Comment by ZangLucarson** [comment]
I just did that, all I found was "libmojoshader.a"

**Comment by ptitSeb** [comment]
oh, a static lib. That will not work here indeed.
In the command line, I think the `-DBUILD_SHARED=ON` didn't stick or something wrong happened.
Command should be 
```
cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUILD_SHARED=ON -DDEPTH_CLIPPING=ON -DFLIP_VIEWPORT=ON
```

**Comment by ZangLucarson** [comment]
Hmm.. I tried and it still can't find "libmojoshader.so"
Is there anything you can think of that I should look for that may be causing the issue?
Also, thank you for the help on this, I really appreciate it :)

**Comment by ZangLucarson** [comment]
I did find this error, could this have anything to do with it? 
`[ 94%] Linking C executable testoutput
/usr/bin/ld: cannot find -lobjc
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/testoutput.dir/build.make:85: testoutput] Error 1
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/testoutput.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[100%] Linking C executable testparse
/usr/bin/ld: cannot find -lobjc
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/testparse.dir/build.make:85: testparse] Error 1
make[1]: *** [CMakeFiles/Makefile2:147: CMakeFiles/testparse.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
`

**Comment by ptitSeb** [comment]
I chacked latest version of mojoshader, and yeah, it has changed a bit. The shader lib flag is now called `BUILD_SHARED_LIBS`
So try
```
cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUILD_SHARED_LIBS=ON -DDEPTH_CLIPPING=ON -DFLIP_VIEWPORT=ON -DPROFILE_METAL=OFF -DPROFILE_GLSPIRV=OFF -DPROFILE_SPIRV=OFF
```

**Comment by ZangLucarson** [comment]
That worked! Thank you so much for the help!

**Comment by ptitSeb** [comment]
Good :)

(I'm curious, what are you try to play, and on what platform?)

**Comment by ZangLucarson** [comment]
I'm trying to play Stardew Valley on the Raspberry Pi 4, I was led here by PI LAB.

**Comment by ptitSeb** [comment]
I see :) (nice videos there !). It should work indeed on the Pi4, and it should be smooth and all !

**Comment by ZangLucarson** [comment]
Alright thank you! There is two more issues I've just run into, one seems to be that TheoraPlay-CS is no longer on Github, and the second one is that when I try compiling the XNAFileDialog library using the "make" command, it says that SDL2/SDL.h is missing.

**Comment by ptitSeb** [comment]
Yeah TheoraPlay-CS is gone. I don't remember if it's needed for Stardew Valley (I think not).
For XNAFileDialog, you are missing another header: `sudo apt install libsdl2-dev` should fix it.

**Comment by ZangLucarson** [comment]
Dang..
Alright, the command for XNAFileDialog worked, fingers crossed I don't run into anything else. Thank you again for the help :)

**Comment by ZangLucarson** [comment]
Does Theorafile require LibTheoraPlay to be compiled, the make command didn't work for it. If not, is it required for Stardew Valley?

**Comment by ptitSeb** [comment]
I checked quickly on my side, I don't think any "theory" is needed for stardew valley.

But I did needed to link some of the system libs with special name (not sure it's needed):
`libSDL2.so` for `libSDL2-2.0.so.1`  and `libsoft_oal.so` for `libopenal.so.1` .

**Comment by ZangLucarson** [comment]
So are these the only libs required for Stardew Valley?

**Comment by ptitSeb** [comment]
try it, you will see if some libs are missing (you need ogg / vorbis / vorbisfile also, at least)

**Comment by ZangLucarson** [comment]
Alright, I will. Thank you.

**Comment by ptitSeb** [comment]
if a native lib is missing, mono will crash and the (large) stack dump will give some hint on what library is missing.
If this happens, put the log here and I'll try to help.

**Comment by ZangLucarson** [comment]
Sounds good, you've been an amazing help :)

**Comment by ZangLucarson** [comment]
So I'm up to the point where I'm putting the game on the pi. I only have the steam copy, but I noticed on the tutorial you mentioned that it would still work with the steam version. Is that correct? I know that the steam copies of the game have steamworks, do I need to do something to get around that?

**Comment by ptitSeb** [comment]
Nothing specific to do, just copy all the files, and then remove the mono dll as indicated.

**Comment by ZangLucarson** [comment]
Which specific mono.dll do I remove? I have 3 different ones: "Mono.Posix.dll","Mono.Security.dll" and "MonoGame.Framework.dll"

**Comment by ptitSeb** [comment]
It's written on the tuto:
```
rm mscorlib.dll
rm System.*dll
rm Mono.*.dll
```

**Comment by ZangLucarson** [comment]
ah, I didn't notice that before, thank you.

**Comment by ZangLucarson** [comment]
Good news! I got it working! Thank you so much @ptitSeb for all of your help!
Bad news! The scaling on the ui is messed up, I think it's just a personal issue that I might be able to smooth over by changing some resolution settings.

**Comment by ptitSeb** [comment]
I noticed you need at least 1280x720 screen res to have a proprer main menu and all...

Good luck and enjoy your play on the Pi4 :D

**Comment by ZangLucarson** [comment]
Thank you!

**Comment by daitangio** [comment]
I have the same issue. I fixed removing some caches from the root dir and repeating the cmake command:
```bash
rm -r CMakeFiles/ CMakeCache.txt
cd build && cmake ..
```
